Fix account ownership check

godexsoft opened the pull request:
Fixes #222 

When recreating the cursor from a marker in `account_objects` we call out to 'isOwnedByAccount' which seems to be a safety net of sorts. 
If the marker happens to land on an entry where the `accountID` we check against is the `destination` the check will fail.
The fix just adds a check for this case. 

On a second thought tho, maybe we don't need this `isOwnedByAccount` check at all?

cjcobb23 commented:
So I think we can probably remove this ownership check. All we are doing is trying to make sure the marker is a valid point to start iteration. I think we can do that check in here: https://github.com/XRPLF/clio/blob/8cc2de5643004fc53bafab328257b2341b73cfd7/src/rpc/RPCHelpers.cpp#L712
and it would be much less likely to be buggy. Specifically, if `hintDir` is not found, or if it is but `hexMarker` is not found in `sfIndexes`, return an error.

godexsoft commented:
> Specifically, if `hintDir` is not found, or if it is but `hexMarker` is not found in `sfIndexes`, return an error.

Take a look if it's what you wanted. I was not sure what to put as the error strings so if you have better suggestions - let me know.

cjcobb23 reviewed (commented):
What is rippled's behavior when you pass an invalid marker?

godexsoft commented:
> What is rippled's behavior when you pass an invalid marker?

@cjcobb23 This is what i got sending an invalid marker:
```
{
  "result": {
    "error": "invalidParams",
    "error_code": 31,
    "error_message": "Invalid field 'marker'.",
    "request": {
      "account": "raeR1nAGkMm18hHpUz1ZP6wGKLA4WCh8rX",
      "command": "account_objects",
      "deletion_blockers_only": false,
      "limit": 10,
      "marker": "1111",
      "validated": true
    },
    "status": "error"
  }
}
```

If i use a valid marker but change one character in it i get this:
```
{
  "result": {
    "account": "raeR1nAGkMm18hHpUz1ZP6wGKLA4WCh8rX",
    "account_objects": [],
    "ledger_current_index": 23293951,
    "status": "success",
    "validated": false
  }
}
```
For the same type of change, with `clio` i get this:
```
{
  "result": {
    "error": "invalidParams",
    "error_code": 31,
    "error_message": "Malformed cursor",
    "status": "error",
    "type": "response",
    "request": {
      "method": "account_objects",
      "params": [
        {
          "marker": "17D55F149C997B7A15F683F9C568A2087DCEAD913962689F9E0A7A56E5F423C5,0",
          "account": "raeR1nAGkMm18hHpUz1ZP6wGKLA4WCh8rX",
          "deletion_blockers_only": false,
          "limit": 10,
          "validated": true
        }
      ]
    }
  },
  "warnings": [
    {
      "id": 2001,
      "message": "This is a clio server. clio only serves validated data. If you want to talk to rippled, include 'ledger_index':'current' in your request"
    }
  ]
}
```
I also get exactly the same error for any other completely malformed marker.

So i guess the takeaway is that we need to distinguish between invalid format of a marker and a marker that does not yield a cursor, right?

godexsoft committed:
Implement account ownership check and paging

Fix account ownership check

[FOLD] Apply changes suggested in review

[FOLD] Implement workaround for empty account_objects

[FOLD] Return empty resultset for account_objects with bogus marker

[FOLD] Patch up double return due to clang-format bug

[FOLD] Put back ledger_index in response as per docs

cjcobb23 commented:
> > What is rippled's behavior when you pass an invalid marker?
> 
> @cjcobb23 This is what i got sending an invalid marker:
> 
> ```
> {
>   "result": {
>     "error": "invalidParams",
>     "error_code": 31,
>     "error_message": "Invalid field 'marker'.",
>     "request": {
>       "account": "raeR1nAGkMm18hHpUz1ZP6wGKLA4WCh8rX",
>       "command": "account_objects",
>       "deletion_blockers_only": false,
>       "limit": 10,
>       "marker": "1111",
>       "validated": true
>     },
>     "status": "error"
>   }
> }
> ```
> 
> If i use a valid marker but change one character in it i get this:
> 
> ```
> {
>   "result": {
>     "account": "raeR1nAGkMm18hHpUz1ZP6wGKLA4WCh8rX",
>     "account_objects": [],
>     "ledger_current_index": 23293951,
>     "status": "success",
>     "validated": false
>   }
> }
> ```
> 
> For the same type of change, with `clio` i get this:
> 
> ```
> {
>   "result": {
>     "error": "invalidParams",
>     "error_code": 31,
>     "error_message": "Malformed cursor",
>     "status": "error",
>     "type": "response",
>     "request": {
>       "method": "account_objects",
>       "params": [
>         {
>           "marker": "17D55F149C997B7A15F683F9C568A2087DCEAD913962689F9E0A7A56E5F423C5,0",
>           "account": "raeR1nAGkMm18hHpUz1ZP6wGKLA4WCh8rX",
>           "deletion_blockers_only": false,
>           "limit": 10,
>           "validated": true
>         }
>       ]
>     }
>   },
>   "warnings": [
>     {
>       "id": 2001,
>       "message": "This is a clio server. clio only serves validated data. If you want to talk to rippled, include 'ledger_index':'current' in your request"
>     }
>   ]
> }
> ```
> 
> I also get exactly the same error for any other completely malformed marker.
> 
> So i guess the takeaway is that we need to distinguish between invalid format of a marker and a marker that does not yield a cursor, right?

Yeah one is malformed, and one is sort of invalid but well formed at least. It looks like rippled is not returning an error for the invalid case, just returning nothing. So we should just copy that.

godexsoft commented on src/rpc/Errors.h:
```diff
@@ -17,6 +17,7 @@ namespace RPC {
  */
 enum class ClioError {
     rpcMALFORMED_CURRENCY = 5000,
+    rpcMARKER_NOT_FOUND = 5001,
```
I'm now thinking whether we should use some rippled code instead of adding our own for this hack.. maybe `rpcOBJECT_NOT_FOUND`?

cjcobb23 commented on src/rpc/RPCHelpers.cpp:
```diff
@@ -686,7 +680,7 @@ traverseOwnedNodes(
             std::end(indexes))
         {
-            return Status(ripple::rpcINVALID_PARAMS, "Marker not found");
+            return Status(ClioError::rpcMARKER_NOT_FOUND);
```
so rippled just returns an empty list in this case right? If so, I'd just return `AccountCursor({beast::zero,0})`. In the handlers, if we return a 0 cursor, we don't return anything to the client

godexsoft replied:
Ah, that might be a really good way if it works. Let me try that

godexsoft replied:
The only thing is that i'm not sure we can handle `validated=false` that way.

cjcobb23 replied:
what is `validated=false`?

godexsoft replied:
If i use a valid marker but change one character in it i get this from rippled:

```
{
  "result": {
    "account": "raeR1nAGkMm18hHpUz1ZP6wGKLA4WCh8rX",
    "account_objects": [],
    "ledger_current_index": 23293951,
    "status": "success",
    "validated": false
  }
}
```

with your suggestion clio now gives:
```
{
  "result": {
    "account": "raeR1nAGkMm18hHpUz1ZP6wGKLA4WCh8rX",
    "account_objects": [],
    "ledger_hash": "5CF222090AFB48C88FB27FD0BB9660404C39E2BFE90B47D8D2E4C5EA7D49598C",
    "ledger_current_index": 23294159,
    "validated": true,
    "status": "success"
  },
  "warnings": [
    {
      "id": 2001,
      "message": "This is a clio server. clio only serves validated data. If you want to talk to rippled, include 'ledger_index':'current' in your request"
    },
    {
      "id": 2002,
      "message": "This server may be out of date"
    }
  ]
}
```

as you can see, rippled sets `validated` to false while `clio` does not. 
If you look at my code changes in this pr you will see that i have added a way for any rpc to set `validated` and not have it be reset to `true` which is literally hardcoded atm. :) 
So to sum up - we need to handle `validated` and set it to `false` in this case.. somehow. With your idea however i can't distinguish empty results vs. this marker that was not found situation.

godexsoft replied:
I could do something like this:

```
    if (auto status = std::get_if<RPC::Status>(&next))
        return *status;

    auto const& nextMarker = std::get<RPC::AccountCursor>(next);
    if (nextMarker.isNonZero())
        response[JS(marker)] = nextMarker.toString();
    else
        response[JS(validated)] = false;

    return response;
```

but this leads to the final paged portion being returned with `validated=false` because apparently the `zero` nextMarker is used to signify end of paged data.

cjcobb23 replied:
oh don't worry about that, thats an intended discrepancy between clio and rippled. clio by default returns validated data, rippled by default returns not yet validated data. If you set `"ledger_index":"validated"` in your request to rippled, this discrepancy won't exist.

godexsoft replied:
I'm only worried about it because i bet that the next bug report will be about this field differ from rippled's output..
Also even for `clio` i think it makes sense that for a bogus marker we get non-validated data (if not an error like before). But i can't see an easier solution to my custom error code approach. Maybe we can ignore it for the time being to move on with the bugs and later deal with the next bug that will be reported for this difference.

cjcobb23 replied:
its ok, everything coming back from clio is always validated data. It can't not be, because clio only stores validated data. The only way to get not validated data from clio is if clio forwards the request to rippled. When you send a request to rippled, you should always set `"ledger_index":"validated"` before comparing to clio's response. Or set it to a number. In your example request, you set `"ledger_current_index"` instead of `"ledger_index"`

cjcobb23 replied:
We have been very clear to everyone from the beginning that clio only serves validated data. That won't change. It's a known and intended discrepancy. It's bad UX to serve not yet validated data

godexsoft replied:
Oh, that's useful to know about `ledger_current_index` not being the correct field to send here. @ManiMounikaKunasani you may have to update your tests with the correct parameter, can you check please?

godexsoft replied:
nvm. re above - i must have added the wrong parameter by mistake while adjusting some unrelated curl request i had in history. I'll remember the difference, thanks.

godexsoft commented on src/rpc/handlers/Subscribe.cpp:
```diff
@@ -442,8 +442,7 @@ doUnsubscribe(Context const& context)
 
-    boost::json::object response = {{"status", "success"}};
-    return response;
+    return boost::json::object{};
```
This is for issue #354. We did not remove this `status` back then but it seems like that's what is needed to match `clio`'s `subscribe` behaviour and `rippled` as well.

cjcobb23 commented on src/rpc/RPCHelpers.cpp:
```diff
@@ -709,22 +670,21 @@ traverseOwnedNodes(
-        {
-            ripple::SerialIter it{hintDir->data(), hintDir->size()};
-            ripple::SLE sle{it, hintIndex.key};
+        if (!hintDir)
+            return Status(ripple::rpcINVALID_PARAMS, "Invalid marker");
```
Is this rippled's behavior? I think this corresponds to passing a marker that has a bogus hint, which is the single integer at the end of the marker.

godexsoft replied:
I think so, yes.

cjcobb23 reviewed (dismissed):
LGTM
